#### Learned:
- .loc[]
- .idxmax()
- .where(df['column'] > 0, 0)
- df['column'] = pd.to_datetime(df['column'])
- df['new_column'] = df['column'].dt.day_name()
- df['column'].value_counts()

# Assignment 3 - Pandas DataFrame
## Part 1 (75 Points evenly distributed to first three questions)
The following code loads the olympics dataset (olympics.csv), which was derived from the Wikipedia entry on [All Time Olympic Games Medals](https://en.wikipedia.org/wiki/All-time_Olympic_Games_medal_table), and does some basic data cleaning. 

The columns are organized as # of Summer games, Summer medals, # of Winter games, Winter medals, total # number of games, total # of medals. Use this dataset to answer the questions below.

In [1]:
import pandas as pd

df = pd.read_csv('olympics.csv', index_col=0, skiprows=1)

for col in df.columns:
    if col[:2]=='01':
        df.rename(columns={col:'Gold'+col[4:]}, inplace=True)
    if col[:2]=='02':
        df.rename(columns={col:'Silver'+col[4:]}, inplace=True)
    if col[:2]=='03':
        df.rename(columns={col:'Bronze'+col[4:]}, inplace=True)
    if col[:1]=='№':
        df.rename(columns={col:'#'+col[1:]}, inplace=True)

names_ids = df.index.str.split('\s\(') # split the index by '('

df.index = names_ids.str[0] # the [0] element is the country name (new index) 
df['ID'] = names_ids.str[1].str[:3] # the [1] element is the abbreviation or ID (take first 3 characters from that)

df = df.drop('Totals')
df.head()

# Summer  Gold  Silver  Bronze  Total  # Winter  Gold.1  \
Afghanistan        13     0       0       2      2         0       0   
Algeria            12     5       2       8     15         3       0   
Argentina          23    18      24      28     70        18       0   
Armenia             5     1       2       9     12         6       0   
Australasia         2     3       4       5     12         0       0   

             Silver.1  Bronze.1  Total.1  # Games  Gold.2  Silver.2  Bronze.2  \
Afghanistan         0         0        0       13       0         0         2   
Algeria             0         0        0       15       5         2         8   
Argentina           0         0        0       41      18        24        28   
Armenia             0         0        0       11       1         2         9   
Australasia         0         0        0        2       3         4         5   

             Combined total   ID  
Afghanistan               2  AFG  
Algeria                  15  ALG  
Argentina                70  ARG  
Armenia                  12  ARM  
Australasia              12  ANZ

In [21]:
df.loc['Afghanistan']

# Summer           13
Gold                0
Silver              0
Bronze              2
Total               2
# Winter            0
Gold.1              0
Silver.1            0
Bronze.1            0
Total.1             0
# Games            13
Gold.2              0
Silver.2            0
Bronze.2            2
Combined total      2
ID                AFG
Name: Afghanistan, dtype: object

### Question 1 

What is the first country in df?

*This function should return a Series.*

In [16]:
# You should write your whole answer within the function provided. 
def answer_zero():
    # This function should return the row for the first country, which is a Series object.
    return df.loc[list(df.index)[0]]

# You can examine what your function returns by calling it in the cell. 
answer_zero() 

# Summer           13
Gold                0
Silver              0
Bronze              2
Total               2
# Winter            0
Gold.1              0
Silver.1            0
Bronze.1            0
Total.1             0
# Games            13
Gold.2              0
Silver.2            0
Bronze.2            2
Combined total      2
ID                AFG
Name: Afghanistan, dtype: object

### Question 2
In summer games, which nation has won the most gold medals?


*This function should return a single string value.*


In [22]:
def answer_one():
    return df.Gold.idxmax()

answer_one()

'United States'

### Question 3
Which nation had the biggest difference on gold medal counts? (between their summer and winter) 

*This function should return a single string value.*

In [31]:
def answer_one():
    df_c = df.copy()
    df_c['diff'] = abs(df_c['Gold'] - df_c['Gold.1'])
    return df_c['diff'].idxmax()

answer_one()

'United States'

# PART 2
## Question 4 (25 Points for first two questions + 25 Points Bonus for last two questions)

We will look at the publicly available airline data in this question similar to flights.csv that we covered in class practices. However, in this assignment, you are given 6 months of seperate data along with lookup tables for carriers and airports. Please apply data exploration and pre-procesing techniques and provide your answers for the following questions.



Questions: 

    1. What carrier has flown the 1st most number of flights? How many?  
    2. Which airport has the 3rd most delays? 
    3. What is the most popular day of the week to travel?
    4. What is the 1st and 5th most flown route? 

Hints:

1- pd.concat(list)
list=[A,B,C...] e.g. A= pd.read_csv("1.csv",encoding='utf-8')

2-please leverage from pandas dataframe features including 
groupby(...).size()...
groupby(...).sum().sort_values(....)

3- Dont forget to consider cancelled flights

4- Try to create a new column for "route"

Dataset Details: 
Dataset name: On-Time Performance,
Lookup Table: Carrier Lookup,
Lookup Table: Airport Lookup


In [39]:
pd.set_option('display.max_columns', None)

In [33]:
ID = pd.read_csv("L_AIRPORT_ID.csv",encoding='utf-8')

ID

Code                                    Description
0     10001         Afognak Lake, AK: Afognak Lake Airport
1     10003  Granite Mountain, AK: Bear Creek Mining Strip
2     10004                       Lik, AK: Lik Mining Camp
3     10005         Little Squaw, AK: Little Squaw Airport
4     10006                     Kizhuyak, AK: Kizhuyak Bay
...     ...                                            ...
6356  16695             Lafayette, TN: Lafayette Municipal
6357  16696               Smith Bay, AK: Smith Bay Airport
6358  16697                Moneta, VA: Smith Mountain Lake
6359  16698               Dune Lake, AK: Dune Lake Airport
6360  99999                        Unknown Point in Alaska

[6361 rows x 2 columns]

In [34]:
HISTORY = pd.read_csv("L_CARRIER_HISTORY.csv",encoding='utf-8')

HISTORY

Code                              Description
0     02Q                  Titan Airways (2006 - )
1     04Q             Tradewind Aviation (2006 - )
2     05Q        Comlux Aviation, AG (2006 - 2012)
3     06Q  Master Top Linhas Aereas Ltd. (2007 - )
4     07Q            Flair Airlines Ltd. (2007 - )
...   ...                                      ...
1841   ZW    Air Wisconsin Airlines Corp (1994 - )
1842   ZX                 Airbc Ltd. (1990 - 2000)
1843   ZX                   Air Georgian (2002 - )
1844   ZY     Atlantic Gulf Airlines (1985 - 1986)
1845  ZYZ       Skyway Aviation Inc. (1960 - 2002)

[1846 rows x 2 columns]

In [40]:
A = pd.read_csv("1.csv",encoding='utf-8')

A

FL_DATE UNIQUE_CARRIER  AIRLINE_ID CARRIER TAIL_NUM  FL_NUM  \
0       2014-01-01             AA       19805      AA   N002AA    2377   
1       2014-01-01             AA       19805      AA   N3FXAA    2378   
2       2014-01-01             EV       20366      EV   N906EV    2500   
3       2014-01-01             EV       20366      EV   N903EV    2502   
4       2014-01-01             EV       20366      EV   N903EV    2502   
...            ...            ...         ...     ...      ...     ...   
471944  2014-01-31             OO       20304      OO   N292SW    5313   
471945  2014-01-31             OO       20304      OO   N580SW    5314   
471946  2014-01-31             OO       20304      OO   N580SW    5314   
471947  2014-01-31             OO       20304      OO   N216SW    5315   
471948  2014-01-31             OO       20304      OO   N216SW    5315   

        ORIGIN_AIRPORT_ID  ORIGIN_AIRPORT_SEQ_ID  ORIGIN_CITY_MARKET_ID  \
0                   12278                1227802                  30928   
1                   13303                1330303                  32467   
2                   11298                1129803                  30194   
3                   11146                1114603                  31146   
4                   11298                1129803                  30194   
...                   ...                    ...                    ...   
471944              12889                1288903                  32211   
471945              12892                1289203                  32575   
471946              14689                1468902                  34689   
471947              11292                1129202                  30325   
471948              14543                1454302                  34543   

       ORIGIN       ORIGIN_CITY_NAME ORIGIN_STATE_ABR  ORIGIN_STATE_FIPS  \
0         ICT            Wichita, KS               KS                 20   
1         MIA              Miami, FL               FL                 12   
2         DFW  Dallas/Fort Worth, TX               TX                 48   
3         CRW  Charleston/Dunbar, WV               WV                 54   
4         DFW  Dallas/Fort Worth, TX               TX                 48   
...       ...                    ...              ...                ...   
471944    LAS          Las Vegas, NV               NV                 32   
471945    LAX        Los Angeles, CA               CA                  6   
471946    SBA      Santa Barbara, CA               CA                  6   
471947    DEN             Denver, CO               CO                  8   
471948    RKS       Rock Springs, WY               WY                 56   

       ORIGIN_STATE_NM  ORIGIN_WAC  DEST_AIRPORT_ID  DEST_AIRPORT_SEQ_ID  \
0               Kansas          62            11298              1129803   
1              Florida          33            15304              1530402   
2                Texas          74            12191              1219102   
3        West Virginia          39            11298              1129803   
4                Texas          74            11146              1114603   
...                ...         ...              ...                  ...   
471944          Nevada          85            11638              1163805   
471945      California          91            14689              1468902   
471946      California          91            12892              1289203   
471947        Colorado          82            14543              1454302   
471948         Wyoming          88            11292              1129202   

        DEST_CITY_MARKET_ID DEST         DEST_CITY_NAME DEST_STATE_ABR  \
0                     30194  DFW  Dallas/Fort Worth, TX             TX   
1                     33195  TPA              Tampa, FL             FL   
2                     31453  HOU            Houston, TX             TX   
3                     30194  DFW  Dallas/Fort Worth, TX             TX   
4                     311

In [41]:
B = pd.read_csv("2.csv",encoding='utf-8')

B

FL_DATE UNIQUE_CARRIER  AIRLINE_ID CARRIER TAIL_NUM  FL_NUM  \
0       2014-02-01             EV       20366      EV   N908EV    2537   
1       2014-02-01             EV       20366      EV   N904EV    2538   
2       2014-02-01             EV       20366      EV   N904EV    2538   
3       2014-02-01             EV       20366      EV   N876AS    2539   
4       2014-02-01             EV       20366      EV   N18557    3810   
...            ...            ...         ...     ...      ...     ...   
430597  2014-02-28             AA       19805      AA   N3KKAA    1599   
430598  2014-02-28             AA       19805      AA   N493AA    1600   
430599  2014-02-28             AA       19805      AA   N493AA    1600   
430600  2014-02-28             AA       19805      AA   N437AA    1602   
430601  2014-02-28             AA       19805      AA   N437AA    1602   

        ORIGIN_AIRPORT_ID  ORIGIN_AIRPORT_SEQ_ID  ORIGIN_CITY_MARKET_ID  \
0                   12191                1219102                  31453   
1                   11298                1129803                  30194   
2                   14814                1481402                  30476   
3                   11298                1129803                  30194   
4                   12266                1226603                  31453   
...                   ...                    ...                    ...   
430597              13303                1330303                  32467   
430598              11298                1129803                  30194   
430599              13851                1385103                  33851   
430600              11298                1129803                  30194   
430601              14893                1489302                  33192   

       ORIGIN       ORIGIN_CITY_NAME ORIGIN_STATE_ABR  ORIGIN_STATE_FIPS  \
0         HOU            Houston, TX               TX                 48   
1         DFW  Dallas/Fort Worth, TX               TX                 48   
2         SHV         Shreveport, LA               LA                 22   
3         DFW  Dallas/Fort Worth, TX               TX                 48   
4         IAH            Houston, TX               TX                 48   
...       ...                    ...              ...                ...   
430597    MIA              Miami, FL               FL                 12   
430598    DFW  Dallas/Fort Worth, TX               TX                 48   
430599    OKC      Oklahoma City, OK               OK                 40   
430600    DFW  Dallas/Fort Worth, TX               TX                 48   
430601    SMF         Sacramento, CA               CA                  6   

       ORIGIN_STATE_NM  ORIGIN_WAC  DEST_AIRPORT_ID  DEST_AIRPORT_SEQ_ID  \
0                Texas          74            11298              1129803   
1                Texas          74            14814              1481402   
2            Louisiana          72            11298              1129803   
3                Texas          74            10747              1074702   
4                Texas          74            14683              1468303   
...                ...         ...              ...                  ...   
430597         Florida          33            12953              1295302   
430598           Texas          74            13851              1385103   
430599        Oklahoma          73            11298              1129803   
430600           Texas          74            14893              1489302   
430601      California          91            11298              1129803   

        DEST_CITY_MARKET_ID DEST         DEST_CITY_NAME DEST_STATE_ABR  \
0                     30194  DFW  Dallas/Fort Worth, TX             TX   
1                     30476  SHV         Shreveport, LA             LA   
2                     30194  DFW  Dallas/Fort Worth, TX             TX   
3                     30747  BRO        Brownsville, TX             TX   
4                     332

In [37]:
C = pd.read_csv("3.csv",encoding='utf-8')

D = pd.read_csv("4.csv",encoding='utf-8')

E = pd.read_csv("5.csv",encoding='utf-8')

F = pd.read_csv("6.csv",encoding='utf-8')


In [38]:
print(ID.shape, HISTORY.shape, A.shape, B.shape, C.shape, D.shape, E.shape, F.shape)

(6361, 2) (1846, 2) (471949, 51) (430602, 51) (503758, 51) (483499, 51) (499278, 51) (502617, 51)


In [44]:
list_of_months = [A,B,C,D,E,F]
flight = pd.concat(list_of_months)

In [45]:
flight.shape

(2891703, 51)

### 1. What carrier has flown the 1st most number of flights? How many?  

In [47]:
flight.CARRIER.value_counts()

WN    579338
DL    383526
EV    350442
OO    304989
AA    267766
UA    242908
US    207776
MQ    202504
B6    123566
AS     76734
FL     49774
F9     37794
HA     36042
VX     28544
Name: CARRIER, dtype: int64

In [52]:
HISTORY.loc[HISTORY['Code'] == 'WN']

Code                       Description
1712   WN  Southwest Airlines Co. (1979 - )

#### Answer: Southwest Airlines Co. has flown the 1st most number of flights: 579338

### 2. Which airport has the 3rd most delays? 


#### Assumptions
1. Flights that came in early should not reduce the total counts of delay
2. I use delay group instead of actual delay time because I think it's a good representation of delay situation

In [66]:
# First, we create extra columns to filter out flights that come early
# Because when we sum up the delay time, these negative values would decrease the total delay time
# (flights that came in early should not reduce the total counts of delay)
# This would lead to an inaccurate estimate of total delay time

flight['departure_delay_without_negative'] = flight['DEP_DELAY_GROUP'].where(flight['DEP_DELAY_GROUP'] > 0, 0)
flight['arrival_delay_without_negative'] = flight['ARR_DELAY_GROUP'].where(flight['ARR_DELAY_GROUP'] > 0, 0)
flight

FL_DATE UNIQUE_CARRIER  AIRLINE_ID CARRIER TAIL_NUM  FL_NUM  \
0       2014-01-01             AA       19805      AA   N002AA    2377   
1       2014-01-01             AA       19805      AA   N3FXAA    2378   
2       2014-01-01             EV       20366      EV   N906EV    2500   
3       2014-01-01             EV       20366      EV   N903EV    2502   
4       2014-01-01             EV       20366      EV   N903EV    2502   
...            ...            ...         ...     ...      ...     ...   
502612  2014-06-30             AA       19805      AA   N5DVAA    2496   
502613  2014-06-30             AA       19805      AA   N5CPAA    2497   
502614  2014-06-30             AA       19805      AA   N202AA    2498   
502615  2014-06-30             AA       19805      AA   N202AA    2498   
502616  2014-06-30             AA       19805      AA   N3DCAA    2499   

        ORIGIN_AIRPORT_ID  ORIGIN_AIRPORT_SEQ_ID  ORIGIN_CITY_MARKET_ID  \
0                   12278                1227802                  30928   
1                   13303                1330303                  32467   
2                   11298                1129803                  30194   
3                   11146                1114603                  31146   
4                   11298                1129803                  30194   
...                   ...                    ...                    ...   
502612              11298                1129803                  30194   
502613              11298                1129803                  30194   
502614              11292                1129202                  30325   
502615              11298                1129803                  30194   
502616              11298                1129803                  30194   

       ORIGIN       ORIGIN_CITY_NAME ORIGIN_STATE_ABR  ORIGIN_STATE_FIPS  \
0         ICT            Wichita, KS               KS                 20   
1         MIA              Miami, FL               FL                 12   
2         DFW  Dallas/Fort Worth, TX               TX                 48   
3         CRW  Charleston/Dunbar, WV               WV                 54   
4         DFW  Dallas/Fort Worth, TX               TX                 48   
...       ...                    ...              ...                ...   
502612    DFW  Dallas/Fort Worth, TX               TX                 48   
502613    DFW  Dallas/Fort Worth, TX               TX                 48   
502614    DEN             Denver, CO               CO                  8   
502615    DFW  Dallas/Fort Worth, TX               TX                 48   
502616    DFW  Dallas/Fort Worth, TX               TX                 48   

       ORIGIN_STATE_NM  ORIGIN_WAC  DEST_AIRPORT_ID  DEST_AIRPORT_SEQ_ID  \
0               Kansas          62            11298              1129803   
1              Florida          33            15304              1530402   
2                Texas          74            12191              1219102   
3        West Virginia          39            11298              1129803   
4                Texas          74            11146              1114603   
...                ...         ...              ...                  ...   
502612           Texas          74            12892              1289203   
502613           Texas          74            12892              1289203   
502614        Colorado          82            11298              1129803   
502615           Texas          74            11292              1129202   
502616           Texas          74            12892              1289203   

        DEST_CITY_MARKET_ID DEST         DEST_CITY_NAME DEST_STATE_ABR  \
0                     30194  DFW  Dallas/Fort Worth, TX             TX   
1                     33195  TPA              Tampa, FL             FL   
2                     31453  HOU            Houston, TX             TX   
3                     30194  DFW  Dallas/Fort Worth, TX             TX   
4                     311

In [74]:
# Departure delay 

departure_delay = flight.groupby('ORIGIN_AIRPORT_ID')['departure_delay_without_negative'].sum()
departure_delay

ORIGIN_AIRPORT_ID
10135     557.0
10136     899.0
10140    8614.0
10141     101.0
10146     250.0
          ...  
15841     185.0
15897      23.0
15919    4812.0
15991     144.0
16218     445.0
Name: departure_delay_without_negative, Length: 319, dtype: float64

In [75]:
# Arrival delay

arrival_delay = flight.groupby('DEST_AIRPORT_ID')['arrival_delay_without_negative'].sum()
arrival_delay

DEST_AIRPORT_ID
10135     544.0
10136     911.0
10140    9729.0
10141     149.0
10146     300.0
          ...  
15841     174.0
15897       4.0
15919    4943.0
15991      89.0
16218     433.0
Name: arrival_delay_without_negative, Length: 319, dtype: float64

In [85]:
# Merged above two dataframes

total_delay = pd.concat([departure_delay, arrival_delay], axis = 1)
total_delay['total_delay'] = total_delay['departure_delay_without_negative'] + total_delay['arrival_delay_without_negative']
total_delay.sort_values(by = 'total_delay', ascending = False).head(3)

departure_delay_without_negative  arrival_delay_without_negative  \
13930                          142777.0                        141275.0   
10397                          118276.0                        114999.0   
11298                           94073.0                         93863.0   

       total_delay  
13930     284052.0  
10397     233275.0  
11298     187936.0

In [158]:
ID.loc[ID['Code'] == 11298]['Description'].values[0]

'Dallas/Fort Worth, TX: Dallas/Fort Worth International'

#### Answer: Dallas/Fort Worth, TX: Dallas/Fort Worth Interational Airport has the 3rd most delays

With 2 assumptions:
- Flights that came in early should not reduce the total counts of delay
- I use delay group instead of actual delay time because I think it's a good representation of delay situation

### 3. What is the most popular day of the week to travel?

In [109]:
flight['FL_DATE'] = pd.to_datetime(flight['FL_DATE'])
flight['day_of_the_week'] = flight['FL_DATE'].dt.day_name()

flight['day_of_the_week'].value_counts()

Friday       437210
Thursday     436422
Monday       433913
Wednesday    424036
Sunday       405367
Tuesday      405063
Saturday     349692
Name: day_of_the_week, dtype: int64

#### Answer: Friday is the most popular day of the week to travel

### 4. What is the 1st and 5th most flown route? 

In [120]:
flight['route'] = flight['ORIGIN_AIRPORT_ID']* 1000000 + flight['DEST_AIRPORT_ID']
flight['route']

0         12278011298
1         13303015304
2         11298012191
3         11146011298
4         11298011146
             ...     
502612    11298012892
502613    11298012892
502614    11292011298
502615    11298011292
502616    11298012892
Name: route, Length: 2891703, dtype: int64

In [126]:
result = list(flight['route'].value_counts().head(5).index)
result

[14771012892, 12892014771, 12889012892, 12892012889, 12478012892]

In [137]:
str(result[0])[0:5]

'14771'

In [138]:
# arrival airport ID 
str(result[0])[-5:]

'12892'

In [177]:
def parse_result(list):
    list_result = []
    for x in list:
        origin_code = str(x)[0:5] # departure airport ID
        destination_code = str(x)[-5:] # arrival airport ID 
        
        origin = ID.loc[ID['Code'] == int(origin_code)]['Description'].values[0]
        destination = ID.loc[ID['Code'] == int(destination_code)]['Description'].values[0]
                
        list_result.append((origin, destination))
    
    return(list_result)

#### Answer:

In [187]:
answer = parse_result(result)

for x in range(len(answer)):
    print('The number ' + str(x+1) + ' most flown route is from: \n' + answer[x][0] + '\n to: \n' + answer[x][1] + '\n\n')

The number 1 most flown route is from: 
San Francisco, CA: San Francisco International
 to: 
Los Angeles, CA: Los Angeles International


The number 2 most flown route is from: 
Los Angeles, CA: Los Angeles International
 to: 
San Francisco, CA: San Francisco International


The number 3 most flown route is from: 
Las Vegas, NV: McCarran International
 to: 
Los Angeles, CA: Los Angeles International


The number 4 most flown route is from: 
Los Angeles, CA: Los Angeles International
 to: 
Las Vegas, NV: McCarran International


The number 5 most flown route is from: 
New York, NY: John F. Kennedy International
 to: 
Los Angeles, CA: Los Angeles International


